[PyTorch] Lite interpreter with a backend delegate

iseeyuan opened the pull request:
1. Decouple the auxiliary functions

The auxiliary functions (getSchema and preProcessFunction related) are used in backend.h. They are generic functions with no dependencies on other PyTorch targets, and they are used on both server side and client side. 

However, they were originally put into `backend_details.cpp`. This files is used to implement the actual lowering function `codegen_backend_module`, which couples to the rest of PyTorch. 

To have the auxiliary functions built on client side with lite interpreter, move them to `backend.cpp`. With this decoupling, the client runtime does not need to couple with `codegen_backend_module`, which is only used on server side.

2. Have a macro, DELEGATE_RUNTIME, to prevent the `preprocess` function from being built with the client runtime, since it's not used on client side.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54462 [PyTorch] Lite interpreter with a backend delegate**

Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

iseeyuan committed:
[PyTorch] Lite interpreter with a backend delegate

Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch] Lite interpreter with a backend delegate"


Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

[ghstack-poisoned]

raziel reviewed (commented):
Thanks!

kimishpatel commented on torch/csrc/jit/backends/backend.cpp:
```diff
@@ -0,0 +1,81 @@
+}
+
+namespace {
+std::unordered_map<std::string, BackendPreprocessFunction>&
+backendPreprocessFunctions() {
+  static std::unordered_map<std::string, BackendPreprocessFunction>
+      preprocess_functions;
+  return preprocess_functions;
+}
```
Why is this function moved here? Isnt this something need for AOT compile on server?

kimishpatel commented on torch/csrc/jit/backends/backend.cpp:
```diff
@@ -0,0 +1,81 @@
+}
+} // namespace
+
+bool hasBackendPreprocessFunction(const std::string& name) {
+  return backendPreprocessFunctions().count(name);
+}
+
+void registerBackendPreprocessFunction(
+    const std::string& name,
+    const BackendPreprocessFunction& preprocess) {
+  TORCH_CHECK(
+      !detail::hasBackendPreprocessFunction(name),
+      "Preprocessing function for backend ",
+      name,
+      " is already registered. Ensure that registration is only called once.");
+  detail::backendPreprocessFunctions()[name] = preprocess;
+}
+
+BackendPreprocessFunction getBackendPreprocessFunction(
+    const std::string& name) {
+  TORCH_CHECK(
+      hasBackendPreprocessFunction(name),
+      "Preprocessing function for backend ",
+      name,
+      " is not registered.");
+  return backendPreprocessFunctions()[name];
```
Same for these.

kimishpatel commented on test/cpp/jit/test_backend_compiler_lib.cpp:
```diff
@@ -110,6 +110,15 @@ namespace {
+    const c10::Dict<IValue, IValue>& method_compile_spec) {
+  return method_compile_spec;
+}
+#else
```
Given that preprocess is used only on server side and we have removed preprocess from backend interface, then why should we have even the dummy preprocess if we are compiling delegate runtime?
Would it be possible or even make sense to refactor https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/backends/backend.h#L23?

iseeyuan replied:
@kimishpatel and @SplitInfinity , good questions! I remember I discussed with @raziel before on the functions related to preprocess. There are two options. 1. to keep as is, but pass a dummy function or null on client side. 2. to move it out of backend.h. This PR is on option 1, but I could see the benefit of option 2. 

I could also see one pro of option2, to have backend registration in one line, including the (optional) preprocess. I remember there is also some backward compatibility needs for glow but I could remember wrong. 

For why I didn't use optional here is because the codes of preprocessing is in the same file of other APIs. Symbols in the server side preprocess can not be compiled. One option is to move preprocess into another file, including the registration part. 

@raziel could you confirm if there's any issue to move to option 2?

raziel replied:
The reason to have preprocess in backend.h is because it is used in the registration of a backend.

I think the discussion here is about registration on-host vs on-target, preprocess only exists on-host. 

I see these options:

1. Keep a single registration mechanism via the backend class: adding a constructor of the backend class that doesn't receive a preprocess function (I would prefer that to having an <optional> or null). We would just document such constructor for on-target registration.

2. Split the registration of preprocess and the backend delegation interface: remove the preprocess argument from the backend class constructor, and expose another API for detail::registerBackendPreprocessFunction(). That preprocess specific API/code should then live on a separate file such that we don't have do ifdef stuff (let me know if this doesn't make sense).

The advantage of (1) is that there's a single mechanism for backend registration, which means as a backend there's less chance for mistakes and you only care for linking backend.h. Caveat, which now I see is that we have to do the ugly macro thing because exactly you now bring a lot of baggage into backend.cpp

The advantage of (2) is that it's explicit what each registration is doing: one for interface, another for preprocessing, which makes particular sense for on-device cases where you really care about keeping those 2 worlds (and their dependencies) separate. That way it's explicit for a backend implementation that they're taking 2 separate dependencies depending on what they are building. Caveat is that there's now 2 things a backend will have to do to register their implementations, but in reality they'll have to do that anyway if they care to have a on-host and on-target builds (which I think many will). Plus, I think the extent of the mistakes that could happen is fairly small (forget to register preprocess or the interface, or mismatched backend names for the 2 registrations), but those should be pretty obvious to catch and fix.

So I suggest going with option #2 above.

Thoughts?

kimishpatel replied:
@raziel I agree with option two. Although for ease of usage for backends that do not have distinct function for on-host and on-target, we can just have `detail::registerBackendWithPreprocessFunction` and `detail::registerBackend`. This way for cases like glow where compile and run is together the former API should suffice.

iseeyuan replied:
@raziel and @kimishpatel , cool! I'll go with option 2, to split the preprocess registration from backend, and put both preprocess implementation and registration in a separate source file. 

I don't see any BC issues since all changes are on source code refactor, but will let you know if I find any. 

I also added a test on this PR, which will be updated soon.

kimishpatel reviewed (commented):
Left a couple of comments.

SplitInfinity reviewed (commented):
I'm curious about the answers to @kimishpatel's questions. I would also recommend renaming (e.g. include `python` in the name) or reorganizing (e.g. make `backends/python` directory) the files to make it more obvious which ones have Python dependencies and which ones do not.

iseeyuan commented on torch/csrc/jit/backends/backend.cpp:
```diff
@@ -0,0 +1,81 @@
+
+namespace torch {
+namespace jit {
+namespace detail {
```
Will remove detail namespace

iseeyuan commented on tools/build_variables.bzl:
```diff
@@ -134,6 +134,7 @@ core_sources_full_mobile = [
     "torch/csrc/jit/api/module.cpp",
     "torch/csrc/jit/api/object.cpp",
+    "torch/csrc/jit/backends/backend.cpp",
     "torch/csrc/jit/backends/backend_detail.cpp",
```
The functions are used in backend registration, which is also used in the mobile client. That's why I put it in core_sources_full_mobile, which is used to build mobile runtime. I assume that `_def_unboxed` is used to register functions in both jit and python, so that it's also needed from C++ runtime?

iseeyuan commented:
> I'm curious about the answers to @kimishpatel's questions. I would also recommend renaming (e.g. include `python` in the name) or reorganizing (e.g. make `backends/python` directory) the files to make it more obvious which ones have Python dependencies and which ones do not.

For Kimish's questions, please refer to follow-up comments in test_backend_compiler_lib.cpp. I'm going to pull preprocess out of backend to have a clear boundary.

For python dependencies, I'm thinking that torchbind is for both C++ and Python, so I cannot just move `_def_unboxed` to python only, because it's also needed in C++ client runtime. Please let me know if my understanding is correct.

iseeyuan committed:
Update on "[PyTorch] Lite interpreter with a backend delegate"

1. Decouple the auxiliary functions

The auxiliary functions (getSchema and preProcessFunction related) are used in backend.h. They are generic functions with no dependencies on other PyTorch targets, and they are used on both server side and client side. 

However, they were originally put into `backend_details.cpp`. This files is used to implement the actual lowering function `codegen_backend_module`, which couples to the rest of PyTorch. 

To have the auxiliary functions built on client side with lite interpreter, move them to `backend.cpp`. With this decoupling, the client runtime does not need to couple with `codegen_backend_module`, which is only used on server side.

2. Have a macro, DELEGATE_RUNTIME, to prevent the `preprocess` function from being built with the client runtime, since it's not used on client side.


Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch] Lite interpreter with a backend delegate"

1. Decouple the auxiliary functions

The auxiliary functions (getSchema and preProcessFunction related) are used in backend.h. They are generic functions with no dependencies on other PyTorch targets, and they are used on both server side and client side. 

However, they were originally put into `backend_details.cpp`. This files is used to implement the actual lowering function `codegen_backend_module`, which couples to the rest of PyTorch. 

To have the auxiliary functions built on client side with lite interpreter, move them to `backend.cpp`. With this decoupling, the client runtime does not need to couple with `codegen_backend_module`, which is only used on server side.

2. Have a macro, DELEGATE_RUNTIME, to prevent the `preprocess` function from being built with the client runtime, since it's not used on client side.


Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch] Lite interpreter with a backend delegate"

1. Decouple the auxiliary functions

The auxiliary functions (getSchema and preProcessFunction related) are used in backend.h. They are generic functions with no dependencies on other PyTorch targets, and they are used on both server side and client side. 

However, they were originally put into `backend_details.cpp`. This files is used to implement the actual lowering function `codegen_backend_module`, which couples to the rest of PyTorch. 

To have the auxiliary functions built on client side with lite interpreter, move them to `backend.cpp`. With this decoupling, the client runtime does not need to couple with `codegen_backend_module`, which is only used on server side.

2. Have a macro, DELEGATE_RUNTIME, to prevent the `preprocess` function from being built with the client runtime, since it's not used on client side.


Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch] Lite interpreter with a backend delegate"

1. Decouple the auxiliary functions

The auxiliary functions (getSchema and preProcessFunction related) are used in backend.h. They are generic functions with no dependencies on other PyTorch targets, and they are used on both server side and client side. 

However, they were originally put into `backend_details.cpp`. This files is used to implement the actual lowering function `codegen_backend_module`, which couples to the rest of PyTorch. 

To have the auxiliary functions built on client side with lite interpreter, move them to `backend.cpp`. With this decoupling, the client runtime does not need to couple with `codegen_backend_module`, which is only used on server side.

2. Have a macro, DELEGATE_RUNTIME, to prevent the `preprocess` function from being built with the client runtime, since it's not used on client side.


Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

[ghstack-poisoned]

kimishpatel commented:
> For python dependencies, I'm thinking that torchbind is for both C++ and Python, so I cannot just move `_def_unboxed` to python only, because it's also needed in C++ client runtime. Please let me know if my understanding is correct.

I think this is correct. Torchbind is binding methods to class defined in torchscript which becomes visible in python but the class itself is in torchscript. So class definition must be available in cpp runtime as well if we are loading objects of that type. So we do need custom class registration in cpp.

raziel commented on torch/csrc/jit/backends/backend_preprocess.h:
```diff
@@ -0,0 +1,18 @@
+  std::string backend_name_;
+
+ public:
+  preprocess_register(
```
We don't need a class for this, right?
I think for backend the reason is that it was needed for the torch binding mechanism of torch::class_<>()

Technically we just need to expose detail::registerBackendPreprocessFunction either directly, or through some other namespace.

My suggestion is to just expose it through a a function that calls detail::registerBackendPreprocessFunction, unless there's an advantage to the class?

In terms of the name I'd make it very clear like registerBackendPreprocessing or something like that.

The same for the file name; maybe be should at least name the file more like what we want in the future: backend_compiler_registry.h or something like that.

iseeyuan replied:
Good point to make the name specific! I have changed it to backend_preprocess_register. 

The reason to make it a class is to have it stay in the global space. If I just call the function, registerBackendPreprocessing, in the global space, there is compiler error. The compiler expects it to be a function declaration. I was stuck in compilation error till I found some example of wrapping it to a class. In such a case, the class object "reg" would stay in the binary. Not sure if there's a better solution on this.

raziel reviewed (commented):
Thanks.

Mainly small comments, and the naming/exposing of the preprocess registration.

SplitInfinity commented on torch/csrc/jit/backends/backend_detail.cpp:
```diff
@@ -10,46 +10,6 @@
 namespace torch {
```
`code_template.h` is needed in order to use `CodeTemplate`. `backend_resolver.h` is needed in order to use `loweredModuleResolver()`.

SplitInfinity reviewed (commented):
+1 to @raziel's comments. I suggest also some renaming because the difference between `backend.h` and `backend_detail.h` is not clear anymore (based on the name).

raziel commented:
Also, before I forget, the PR description needs to be updated :-)

iseeyuan commented on test/cpp/lite_interpreter_runtime/test_lite_interpreter_runtime.cpp:
```diff
@@ -49,6 +49,18 @@ TEST(RunTimeTest, LoadAndForward) {
+TEST(RunTimeTest, Delegate) {
+  std::string filePath(__FILE__);
+  auto testModelFile = filePath.substr(0, filePath.find_last_of("/\\") + 1);
+  testModelFile.append("delegate_test.ptl");
```
> It might also be a good idea to include a script to generate it in case it needs to be updated in the future. You could also commit the generation script that and run that first during the test instead of committing the `.ptl`.

Generating the .ptl is covered in test_backend.cpp. There we have the full jit to compile, generate and run the model. In this test environment we want to run it with minimum binary, and to test the existing model with no backward compatibility break.

raziel replied:
Maybe just add a comment indicating how to generate it from test_backend.cpp

Commit:
Update on "[PyTorch] Lite interpreter with a backend delegate"

1. Decouple the auxiliary functions

The auxiliary functions (getSchema and preProcessFunction related) are used in backend.h. They are generic functions with no dependencies on other PyTorch targets, and they are used on both server side and client side. 

However, they were originally put into `backend_details.cpp`. This files is used to implement the actual lowering function `codegen_backend_module`, which couples to the rest of PyTorch. 

To have the auxiliary functions built on client side with lite interpreter, move them to `backend.cpp`. With this decoupling, the client runtime does not need to couple with `codegen_backend_module`, which is only used on server side.

2. Have a macro, DELEGATE_RUNTIME, to prevent the `preprocess` function from being built with the client runtime, since it's not used on client side.


Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

raziel reviewed (approved):
Thanks Martin.

Barring any further comments from @SplitInfinity I think looks good.

SplitInfinity commented:
No further comments from me!

cccclai commented on test/cpp/lite_interpreter_runtime/CMakeLists.txt:
```diff
@@ -6,13 +6,22 @@ set(LITE_INTERPRETER_RUNTIME_TEST_DIR
+        ${TORCH_ROOT}/torch/csrc/jit/backends/backend_interface.cpp
+        )
+target_compile_definitions(backend_with_compiler_runtime)
+target_link_libraries(backend_with_compiler_runtime torch)
```
Is this line needed? Does backend_with_compiler_runtime need libtorch?

iseeyuan replied:
Yes there are some references of IValue, Tensor, etc. Here I assume libtorch is the lite libtorch?

cccclai replied:
Yeah if it's with BUILD_LITE_INTERPRETER ON.

iseeyuan replied:
Added PRIVATE when linking torch to backend_with_compiler_runtime

cccclai commented on test/cpp/lite_interpreter_runtime/CMakeLists.txt:
```diff
@@ -6,13 +6,22 @@ set(LITE_INTERPRETER_RUNTIME_TEST_DIR
+  ${ATen_CPU_INCLUDE}
+)
+
+target_link_libraries(test_lite_interpreter_runtime PRIVATE torch gtest backend_with_compiler_runtime)
```
test_lite_interpreter_runtime is linked with torch again, so I'm a bit confused.

iseeyuan replied:
I could remove torch here but I assume it should be fine, as long as we don't have circular dependency. The reason I keep it here is to make it explicit that test_lite_interpreter_runtime depends on torch. Any potential risk on this way?

cccclai replied:
I see. Probably add a comment so others know why we have it like this?

iseeyuan committed:
Update on "[PyTorch] Lite interpreter with a backend delegate"

1. Decouple the auxiliary functions

The auxiliary functions (getSchema and preProcessFunction related) are used in backend.h. They are generic functions with no dependencies on other PyTorch targets, and they are used on both server side and client side. 

However, they were originally put into `backend_details.cpp`. This files is used to implement the actual lowering function `codegen_backend_module`, which couples to the rest of PyTorch. 

To have the auxiliary functions built on client side with lite interpreter, move them to `backend.cpp`. With this decoupling, the client runtime does not need to couple with `codegen_backend_module`, which is only used on server side.

2. Have a macro, DELEGATE_RUNTIME, to prevent the `preprocess` function from being built with the client runtime, since it's not used on client side.


Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch] Lite interpreter with a backend delegate"

1. Decouple the auxiliary functions

The auxiliary functions (getSchema and preProcessFunction related) are used in backend.h. They are generic functions with no dependencies on other PyTorch targets, and they are used on both server side and client side. 

However, they were originally put into `backend_details.cpp`. This files is used to implement the actual lowering function `codegen_backend_module`, which couples to the rest of PyTorch. 

To have the auxiliary functions built on client side with lite interpreter, move them to `backend.cpp`. With this decoupling, the client runtime does not need to couple with `codegen_backend_module`, which is only used on server side.

2. Have a macro, DELEGATE_RUNTIME, to prevent the `preprocess` function from being built with the client runtime, since it's not used on client side.


Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch] Lite interpreter with a backend delegate"

1. Decouple the auxiliary functions

The auxiliary functions (getSchema and preProcessFunction related) are used in backend.h. They are generic functions with no dependencies on other PyTorch targets, and they are used on both server side and client side. 

However, they were originally put into `backend_details.cpp`. This files is used to implement the actual lowering function `codegen_backend_module`, which couples to the rest of PyTorch. 

To have the auxiliary functions built on client side with lite interpreter, move them to `backend.cpp`. With this decoupling, the client runtime does not need to couple with `codegen_backend_module`, which is only used on server side.

2. Have a macro, DELEGATE_RUNTIME, to prevent the `preprocess` function from being built with the client runtime, since it's not used on client side.


Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch] Lite interpreter with a backend delegate"

1. Decouple the auxiliary functions

The auxiliary functions (getSchema and preProcessFunction related) are used in backend.h. They are generic functions with no dependencies on other PyTorch targets, and they are used on both server side and client side. 

However, they were originally put into `backend_details.cpp`. This files is used to implement the actual lowering function `codegen_backend_module`, which couples to the rest of PyTorch. 

To have the auxiliary functions built on client side with lite interpreter, move them to `backend.cpp`. With this decoupling, the client runtime does not need to couple with `codegen_backend_module`, which is only used on server side.

2. Have a macro, DELEGATE_RUNTIME, to prevent the `preprocess` function from being built with the client runtime, since it's not used on client side.


Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch] Lite interpreter with a backend delegate"

1. Decouple the auxiliary functions

The auxiliary functions (getSchema and preProcessFunction related) are used in backend.h. They are generic functions with no dependencies on other PyTorch targets, and they are used on both server side and client side. 

However, they were originally put into `backend_details.cpp`. This files is used to implement the actual lowering function `codegen_backend_module`, which couples to the rest of PyTorch. 

To have the auxiliary functions built on client side with lite interpreter, move them to `backend.cpp`. With this decoupling, the client runtime does not need to couple with `codegen_backend_module`, which is only used on server side.

2. Have a macro, DELEGATE_RUNTIME, to prevent the `preprocess` function from being built with the client runtime, since it's not used on client side.


Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch] Lite interpreter with a backend delegate"

1. Decouple the auxiliary functions

The auxiliary functions (getSchema and preProcessFunction related) are used in backend.h. They are generic functions with no dependencies on other PyTorch targets, and they are used on both server side and client side. 

However, they were originally put into `backend_details.cpp`. This files is used to implement the actual lowering function `codegen_backend_module`, which couples to the rest of PyTorch. 

To have the auxiliary functions built on client side with lite interpreter, move them to `backend.cpp`. With this decoupling, the client runtime does not need to couple with `codegen_backend_module`, which is only used on server side.

2. Have a macro, DELEGATE_RUNTIME, to prevent the `preprocess` function from being built with the client runtime, since it's not used on client side.


Unclean files during sync - Sat Mar 20 04:00:02 PDT 2021


Unclean files during sync - Sun Mar 21 04:00:01 PDT 2021

Differential Revision: [D27232309](https://our.internmc.facebook.com/intern/diff/D27232309/)

[ghstack-poisoned]
